### ISSUE TITLE
Weights and Biases

### ISSUE THREAD

**Issue by ngupta23** [ngupta23/Data-Science-Knowlege-Base#205]
## Overview
 
* [website](https://wandb.ai/site/)
* [github repo examples](https://github.com/wandb/examples)
 
 
### Overview
* It is the output of a process. In ML, most important artifacts are "data" and "model" (binaries).
* Git is good for code, not for artifacts. That is where W&B can help by storing artifacts.
* These are either the outputs of runs (e.g. models) or the input of runs (e.g. datasets - training, validation, test).
  
### Examples
* [Video](https://www.youtube.com/watch?v=Hd94gatGMic&list=PLD80i8An1OEGajeVo15ohAQYF1Ttle0lk) and [Example Notebook for video](https://github.com/wandb/examples/blob/master/colabs/wandb-artifacts/Pipeline_Versioning_with_W%26B_Artifacts.ipynb)
* [More Example Notebooks](https://github.com/wandb/examples/tree/master/colabs/wandb-artifacts)

#### Logging Data
* Initialize:
  * `wandb.init(project="xyz", job-type="load-data")`
  * job-type for organizing artifacts or pipeline step (each pipeline step could be a job-type)
* Create Atrifact: `wandb.Artifact()  # what are we logging, e.g. metadata`
* Treat artifact as a directory.
  * Add files/data to it using `raw_data.new_file()`
  * `run.log_artifact()`

```python
def load_and_log():

    # 🚀 start a run, with a type to label it and a project it can call home
    with wandb.init(project="artifacts-example", job_type="load-data") as run:
        
        datasets = load()  # separate code for loading the datasets
        names = ["training", "validation", "test"]

        # 🏺 create our Artifact
        raw_data = wandb.Artifact(
            "mnist-raw", type="dataset",
            description="Raw MNIST dataset, split into train/val/test",
            metadata={"source": "torchvision.datasets.MNIST",
                      "sizes": [len(dataset) for dataset in datasets]})

        for name, data in zip(names, datasets):
            # 🐣 Store a new file in the artifact, and write something into its contents.
            with raw_data.new_file(name + ".pt", mode="wb") as file:
                x, y = data.tensors
                torch.save((x, y), file)

        # ✍️ Save the artifact to W&B.
        run.log_artifact(raw_data)

load_and_log()
``` 

* Artifacts are portable across projects. 

#### Using Artifacts

* `run.use_artifact('mnist-raw:latest')`
  * Needs the name of the artifact and the version.
  * The last version of the artifact that was uploaded is labelled `latest` 
* `raw_data_artifact.download()`: makes sure the artifact is present in the local filesystem before we can use it.

```python
def preprocess_and_log(steps):

    with wandb.init(project="artifacts-example", job_type="preprocess-data") as run:

        processed_data = wandb.Artifact(
            "mnist-preprocess", type="dataset",
            description="Preprocessed MNIST dataset",
            metadata=steps)  # logging steps to recreate it as metadata.
         
        # ✔️ declare which artifact we'll be using
        raw_data_artifact = run.use_artifact('mnist-raw:latest')

        # 📥 if need be, download the artifact
        raw_dataset = raw_data_artifact.download()
        
        for split in ["training", "validation", "test"]:
            raw_split = read(raw_dataset, split)
            processed_dataset = preprocess(raw_split, **steps)

            with processed_data.new_file(split + ".pt", mode="wb") as file:
                x, y = processed_dataset.tensors
                torch.save((x, y), file)

        run.log_artifact(processed_data)


def read(data_dir, split):
    filename = split + ".pt"
    x, y = torch.load(os.path.join(data_dir, filename))

    return TensorDataset(x, y)
```

#### Model Training

* always pass parameters of the run as metadata for better visibility later
* Also `artifact_obj.add_file()` is an alternate way to add local files to artifacts. In general, just use `new_file` as it is safer.

```python
def build_model_and_log(config):
    with wandb.init(project="artifacts-example", job_type="initialize", config=config) as run:
        config = wandb.config
        
        model = ConvNet(**config)

        model_artifact = wandb.Artifact(
            "convnet", type="model",
            description="Simple AlexNet style CNN",
            metadata=dict(config))

        torch.save(model.state_dict(), "initialized_model.pth")
        # ➕ another way to add a file to an Artifact
        model_artifact.add_file("initialized_model.pth")

        wandb.save("initialized_model.pth")

        run.log_artifact(model_artifact)

model_config = {"hidden_layer_sizes": [32, 64],
                "kernel_sizes": [3],
                "activation": "ReLU",
                "pool_sizes": [2],
                "dropout": 0.5,
                "num_classes": 10}

build_model_and_log(model_config)
```

 
 

## Hyperparameter Sweep
* [Video](https://www.youtube.com/watch?v=9zrmUIlScdY) and [Example Notebook for video](https://github.com/wandb/examples/blob/master/colabs/pytorch/Organizing_Hyperparameter_Sweeps_in_PyTorch_with_W%26B.ipynb)
* [More Examples](https://github.com/wandb/examples/tree/master/examples/wandb-sweeps)
* 3 steps 
  * Define the Sweep (yaml or dict)  
  * Initialize the sweep with the definition: `sweep_id = wandb.sweep()`   
  * Run the sweep agent: `wandb.agent(sweep_id, function=train, count=5)  # 5 runs for random sweep`
    * Example of how `train` uses the config: `wandb.init()`, `wandb.config()` and `wandb.log()`
    * ![Image](https://github.com/user-attachments/assets/574f3844-6927-445e-8be1-6ff0d72c1076)
  * Results
    * hyperparameter importance obtained by running a random forest (refer video)
      * learning rate is important & correlation is positive (when learning rate increases, loss increases)
      * When optimizer = adam, loss goes up
      * Can also use the parallel coordinate plot
    * ![Image](https://github.com/user-attachments/assets/7d564f9a-4984-4968-ab6d-683c71f87253)
